subsystems/DrivetrainBase.java

```diff
@@ -75,23 +150,202 @@ public void tankDriveVoltsReverse(double leftVolts, double rightVolts) {
         this.tankDriveVolts(-leftVolts, -rightVolts);
     }
 
+    // sets the speeds of the speedcontrollergroups rather than the differentrialdrive
+    public void setDrive(double speed) {
+        leftGroup.set(speed);
+        rightGroup.set(speed);
+    }
+
+    /* PID AND ODOMETRY */
+    // NOTE: many of the following methods are self-explanatory. Only the neccesary documentation will be made.
+    public double getLeftPositionRotation() {
+        return leftLeader.getSelectedSensorPosition();
+    }
+
+    public double getRightPositionRotation() {
+        return rightLeader.getSelectedSensorPosition();
+    }
+
+    private double rotationsToMeters(double rotations) {
+        // number of rotations * circumference of wheel
+        return rotations * constants.wheelDiameterMeters * Math.PI;
+    }
+
+    public double getLeftPositionMeters() {
+        return rotationsToMeters(getLeftPositionRotation());
+    }
+
+    public double getRightPositionMeters() {
+        return rotationsToMeters(getRightPositionRotation());
+    }
+
+    public double getLeftVelocityRPM() {
+        return leftLeader.getSelectedSensorVelocity(); // TODO: Removed / 2. Check if correct.
+    }
+
+    public double getRightVelocityRPM() {
+        return rightLeader.getSelectedSensorVelocity(); // TODO: Removed / 2. Check if correct.
+    }
+
+    // converts rotations per minute (RPM) to meters per second (MPS)
+    private double rpmToMPS(double rotationsPerMinute) {
+        double radiansPerSecond = Units.rotationsPerMinuteToRadiansPerSecond(rotationsPerMinute);
+        double linearMetersPerSecond = radiansPerSecond * (constants.wheelDiameterMeters / 2);
+        return linearMetersPerSecond;
+    }
+
+    public double getLeftVelocityMPS() {
+        return rpmToMPS(getLeftVelocityRPM());
+    }
+
+    public double getRightVelocityMPS() {
+        return rpmToMPS(getRightVelocityRPM());
+    }
+
+    // gets the average velocity of the robot by taking the mean of both sides
+    public double getAbsoluteAverageVelocityMPS() {
+        return (Math.abs(getRightVelocityMPS()) + Math.abs(getLeftVelocityMPS())) / 2;
+    }
+
     // gets the direction of the robot
     public double getHeading() {
-        return ahrs.getAngle() * (invertGyro ? -1 : 1);
-        // return gyro.getAngle();
+        try {
+            return ahrs.getAngle() * (invertGyro ? -1 : 1);
+        }
+        catch(NullPointerException e) {
+            System.out.println("Ur mom is a nullpointerexception");
```

Comment:
who wrote this please :sob: this is public code. make this a little more description of what is happening when the nullpointer is actually getting thrown.

Reply:
all my homies has NullPointerException

Reply:
i genuinely have no idea whether you wrote this or hazel did and tbh it could be either of you and i'd believe you

Reply:
kk bestie they are being handled a bit better now